File: sciho.hs
#!/usr/bin/env runghc

import qualified Data.List as List
import qualified Data.List.Split as Split
import qualified Data.Map.Strict as Map

minRating = 7.5 :: Float 
minVotes = 1000 :: Int

tabSplit = map (\l -> Split.splitOn "\t" l)
nTuple = map (\[i,r,v] -> (i, (read r::Float), (read v :: Int))) 
fStr t y r = (t ++ " (" ++ y ++ "), " ++ show r)

main = do
  rawRatings <- readFile "title.ratings.tsv"
  let ts = tabSplit . drop 1 . lines $ rawRatings
  let rs = Map.fromList [ (i, read r :: Float) | [i,r,v] <- ts,
                                                 r >= show minRating,
                                                 let vi = read v :: Int,
                                                 vi >= minVotes ]
  
  rawBasics <- readFile "title.basics.tsv"
  let bs = tabSplit . drop 1 . lines $ rawBasics
  let bs' = [ (i,ty,ti,y,Split.splitOn "," gs) | [i,ty,ti,_,_,y,_,_,gs] <- bs ]
  let bo = [ fStr ti y r | (i,ty,ti,y,gs) <-  bs',
                             Map.member i rs,
                             "Sci-Fi" `elem` gs,
                             "Horror" `elem` gs || "Thriller" `elem` gs,
                             ty == "movie",
                             let r = (rs Map.! i) ]

  mapM_ putStrLn bo

File: sciho.js
#!/usr/bin/env node

const fs = require("fs")
const readline = require("readline")
const stream = require("stream")

const minVotes = 1000
const minRating = 7.5

let ratings = {}
readline.createInterface(
    fs.createReadStream("title.ratings.tsv"),
    new stream()
).on("line", l => {
    let [tid, r, v] = l.split("\t")
    if (r >= minRating && v >= minVotes) { ratings[tid] = r }
})

readline.createInterface(
    fs.createReadStream("title.basics.tsv"),
    new stream()
).on("line", l => {
    var [tid, type, title, _, _, year, _, _, genre] = l.split("\t")
    let genres = genre.split(",")
    let rating = ratings[tid]

    if (
        type === "movie"
        && genres.includes("Sci-Fi")
        && (genres.includes("Horror") || genres.includes("Thriller"))
        && Object.keys(ratings).includes(tid)
    ) { console.log(`${title} (${year}), ${rating}`) }
})

